File: vivid-app/app/src/main/java/com/vivid/app/domain/repository/AccountRepository.kt
package com.vivid.app.domain.repository

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.vivid.app.data.storage.StorageProvider
import kotlinx.coroutines.tasks.await
import javax.inject.Inject

class AccountRepository @Inject constructor(
    private val firestore: FirebaseFirestore,
    private val auth: FirebaseAuth,
    private val storage: StorageProvider
) {
    suspend fun closeAccountAndPurgeData(targetUserId: String): Boolean {
        if (targetUserId.isBlank()) return false
        return try {
            val postsSnapshot = firestore.collection("posts")
                .whereEqualTo("userId", targetUserId)
                .get()
                .await()
            for (doc in postsSnapshot.documents) {
                val storageKey = doc.getString("storageKey")
                if (!storageKey.isNullOrBlank()) {
                    try { storage.deleteFile(storageKey) } catch (_: Exception) {}
                }
                doc.reference.delete().await()
            }
            firestore.collection("users").document(targetUserId).delete().await()
            true
        } catch (e: Exception) {
            false
        }
    }
}

File: vivid-app/iosApp/iosApp/Services/Firestore/AccountRepository.swift
import FirebaseAuth
import FirebaseFirestore
import Foundation

/// Elimina contenido remoto y sus objetos B2 antes de borrar perfil y usuario
/// de Firebase Auth. Si Auth exige login reciente, el error se muestra al usuario.
final class AccountRepository {
    private let db = Firestore.firestore()
    private let auth = Auth.auth()
    private let storage = MediaStorageRepository()

    func deleteCurrentAccount() async throws {
        guard let user = auth.currentUser else { throw FirebaseRepositoryError.unauthenticated }
        let uid = user.uid
        for collection in ["posts", "reels", "stories"] {
            let snapshot = try await FirebaseAsync.value { completion in self.db.collection(collection).whereField("userId", isEqualTo: uid).getDocuments(completion: completion) }
            for document in snapshot.documents {
                if let key = document.data()["storageKey"] as? String, !key.isEmpty { try? await storage.delete(path: key) }
                try await document.reference.deleteAsync()
            }
        }
        try await db.collection("users").document(uid).deleteAsync()
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            user.delete { error in
                if let error { continuation.resume(throwing: error) } else { continuation.resume() }
            }
        }
    }
}
